### DISCUSSION TITLE
Duplicate page numbers with Document.insert_pdf()

### DISCUSSION THREAD

**Discussion by wabiloo** [pymupdf/PyMuPDF#3346] [Looking for help]
I have used Document.insert_pdf() to insert a single cover PDF page at the beginning of a PDF document.
All working fine.

When opening it in MacOS preview or in Acrobat Reader however, that page and the next both have a page number of "1".

![Screenshot 2024-04-05 at 16 38 34](https://github.com/pymupdf/PyMuPDF/assets/4056263/ec7e6b81-81b9-485d-ab47-2b96ec362983)
![Screenshot 2024-04-05 at 16 39 20](https://github.com/pymupdf/PyMuPDF/assets/4056263/a4eeedf0-1173-4265-bfab-1a2b17c5478f)

Is that normal?  Is there a way to renumber the pages?

Note: this also seems to affect my ability to work with links. Links in the original PDF document after insertion are still pointing to the correct page (ie. clicking on them sends me to the correct page), but if I try to create a TOC with them, I have to add 1 to get the TOC links to go to the correct page.

**Comment by wabiloo** [comment]
Partially answering my own question. 
Those are labels, not page numbers. `Document.get_page_labels()` and `Document.set_page_labels()` are used to get/set them (and prefix them, etc).

However that does not really help with the links/TOC side of things. I find it strange that when defining the TOC from the page locations in the links, I need to add +1 to compensate for the inserted content...

**Reply by wabiloo** [reply]
It looks like it's got nothing to do with the cover page, but the fact that page numbers in TOCs are 1-based, whereas page numbers in links are 0-based.

**Reply by JorjMcKie** [reply]
You do not provide example files - but:
* All PDF viewers display page numbers 1-based - which is the reason why the TOC list generated by `page.get_toc()` also does it.
* There in general is a difference between that page number and the optional page **_label_** in a PDF. The PDF creator has the option to assign labels like Roman numbers or arbitrary other strings to pages. If this is present, then PDF viewers usually also show the page label. This may sometimes be confusing if not focusing on this.
* In a programmatic context, page numbers are normally 0-based because this number serves as an index in an array or list of all pages of a document. If you build a detailed (non-simple) TOC via `doc.get_toc(simple=False)`, then each item looks like `(level, title, pno, link)`. The page number `pno` is 1-based, but the "page" key in the additionally provided `link` dictionary provides the 0-based page number.
